test/integration/controller_instance_test.go

```diff
@@ -1261,43 +1261,43 @@ func TestDeleteServiceInstance(t *testing.T) {
 				}
 			},
 		},
-		{
-			name: "deprovision instance after in progress provision",
-			skipVerifyingInstanceSuccess: true,
-			setup: func(ct *controllerTest) {
-				ct.osbClient.PollLastOperationReaction = &fakeosb.PollLastOperationReaction{
-					Response: &osb.LastOperationResponse{
-						State: osb.StateInProgress,
-					},
-				}
-				ct.osbClient.ProvisionReaction = &fakeosb.ProvisionReaction{
-					Response: &osb.ProvisionResponse{
-						Async: true,
-					},
-				}
-				ct.osbClient.DeprovisionReaction = &fakeosb.DeprovisionReaction{
-					Response: &osb.DeprovisionResponse{},
-				}
-			},
-			testFunction: func(ct *controllerTest) {
-				verifyCondition := v1beta1.ServiceInstanceCondition{
-					Type:   v1beta1.ServiceInstanceConditionReady,
-					Status: v1beta1.ConditionFalse,
-					Reason: "ProvisionRequestInFlight",
-				}
-				if err := util.WaitForInstanceCondition(ct.client, testNamespace, testInstanceName, verifyCondition); err != nil {
-					t.Fatalf("error waiting for instance condition: %v", err)
-				}
-				ct.osbClient.PollLastOperationReaction = &fakeosb.PollLastOperationReaction{
```

Comment:
This `PollLastOperationReaction` being accessed by the thread that is handling the PollLastOperation request in the fake broker. At the same time, it is being modified in the thread running the test.

Reply:
This is precisely why the `DynamicPollLastOperationReaction` and its ilk were adding to the fake broker. The dynamic reactions can use atomic variables that are controlled entirely by the test to change the behavior of the reaction without causing data races.